# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/test/gtest/TestQuotaManager.cpp
# Commit: 9987ebf11f14
# Full Hash: 9987ebf11f141d7e31c958972aee01676d717246
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2025-06-16 09:38:11
# Regressor Bug: 1962133
# File Overlap Count: 14
# Description:
#   Bug 1962133 - QM: Add accessed flag to FullOriginMetadata and persist it in the origin metadata file; r=dom-storage-reviewers,hsingh
#   
#   This patch introduces a new mAccessed field in OriginStateMetadata, which is
#   inherited by FullOriginMetadata. One of the reserved 32-bit fields in the
#   metadata header is now used as a bitfield, with one bit allocated to represent
# ==============================================================================

diff -r d2ec302b7103 -r 9987ebf11f14 dom/quota/test/gtest/TestQuotaManager.cpp
--- a/dom/quota/test/gtest/TestQuotaManager.cpp	Sun Jun 15 21:21:24 2025 +0000
+++ b/dom/quota/test/gtest/TestQuotaManager.cpp	Sun Jun 15 23:24:14 2025 +0000
@@ -416,6 +416,8 @@
 // Test simple OpenClientDirectory behavior and verify that origin access time
 // updates are triggered as expected.
 TEST_F(TestQuotaManager, OpenClientDirectory_Simple) {
+  auto testOriginMetadata = GetTestOriginMetadata();
+
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
@@ -424,6 +426,12 @@
   const auto saveOriginAccessTimeCountInternalBefore =
       SaveOriginAccessTimeCountInternal();
 
+  // Can't check origin state metadata since storage is not yet initialized.
+
+  auto directoryMetadataHeaderBefore =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_FALSE(directoryMetadataHeaderBefore);
+
   PerformOnBackgroundThread([]() {
     QuotaManager* quotaManager = QuotaManager::Get();
     ASSERT_TRUE(quotaManager);
@@ -456,6 +464,15 @@
                 saveOriginAccessTimeCountInternalBefore,
             2u);
 
+  auto originStateMetadataAfter = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataAfter);
+  ASSERT_TRUE(originStateMetadataAfter->mAccessed);
+
+  auto directoryMetadataHeaderAfter =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_TRUE(directoryMetadataHeaderAfter);
+  ASSERT_TRUE(directoryMetadataHeaderAfter->mAccessed);
+
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
 
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
@@ -464,6 +481,8 @@
 // Test simple OpenClientDirectory behavior when the origin directory exists,
 // and verify that access time updates are triggered on first and last access.
 TEST_F(TestQuotaManager, OpenClientDirectory_Simple_OriginDirectoryExists) {
+  auto testOriginMetadata = GetTestOriginMetadata();
+
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
@@ -478,6 +497,15 @@
   const auto saveOriginAccessTimeCountInternalBefore =
       SaveOriginAccessTimeCountInternal();
 
+  auto originStateMetadataBefore = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataBefore);
+  ASSERT_FALSE(originStateMetadataBefore->mAccessed);
+
+  auto directoryMetadataHeaderBefore =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_TRUE(directoryMetadataHeaderBefore);
+  ASSERT_FALSE(directoryMetadataHeaderBefore->mAccessed);
+
   PerformOnBackgroundThread([]() {
     QuotaManager* quotaManager = QuotaManager::Get();
     ASSERT_TRUE(quotaManager);
@@ -510,6 +538,15 @@
                 saveOriginAccessTimeCountInternalBefore,
             2u);
 
+  auto originStateMetadataAfter = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataAfter);
+  ASSERT_TRUE(originStateMetadataAfter->mAccessed);
+
+  auto directoryMetadataHeaderAfter =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_TRUE(directoryMetadataHeaderAfter);
+  ASSERT_TRUE(directoryMetadataHeaderAfter->mAccessed);
+
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
 
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
@@ -520,6 +557,8 @@
 // solely for updating access time.
 TEST_F(TestQuotaManager,
        OpenClientDirectory_Simple_NonExistingOriginDirectory) {
+  auto testOriginMetadata = GetTestOriginMetadata();
+
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
@@ -534,6 +573,14 @@
   const auto saveOriginAccessTimeCountInternalBefore =
       SaveOriginAccessTimeCountInternal();
 
+  auto originStateMetadataBefore = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataBefore);
+  ASSERT_FALSE(originStateMetadataBefore->mAccessed);
+
+  auto directoryMetadataHeaderBefore =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_FALSE(directoryMetadataHeaderBefore);
+
   PerformOnBackgroundThread([]() {
     QuotaManager* quotaManager = QuotaManager::Get();
     ASSERT_TRUE(quotaManager);
@@ -571,6 +618,14 @@
                 saveOriginAccessTimeCountInternalBefore,
             0u);
 
+  auto originStateMetadataAfter = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataAfter);
+  ASSERT_TRUE(originStateMetadataAfter->mAccessed);
+
+  auto directoryMetadataHeaderAfter =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_FALSE(directoryMetadataHeaderAfter);
+
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
 
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
@@ -597,6 +652,15 @@
   const auto saveOriginAccessTimeCountInternalBefore =
       SaveOriginAccessTimeCountInternal();
 
+  auto originStateMetadataBefore = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataBefore);
+  ASSERT_FALSE(originStateMetadataBefore->mAccessed);
+
+  auto directoryMetadataHeaderBefore =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_TRUE(directoryMetadataHeaderBefore);
+  ASSERT_FALSE(directoryMetadataHeaderBefore->mAccessed);
+
   PerformOnBackgroundThread([testOriginMetadata]() {
     QuotaManager* quotaManager = QuotaManager::Get();
     ASSERT_TRUE(quotaManager);
@@ -642,6 +706,13 @@
                 saveOriginAccessTimeCountInternalBefore,
             2u);
 
+  auto originStateMetadataAfter = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_FALSE(originStateMetadataAfter);
+
+  auto directoryMetadataHeaderAfter =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_FALSE(directoryMetadataHeaderAfter);
+
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
 
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
@@ -652,6 +723,8 @@
 // after the origin access time update triggered by first access has finished,
 // and that access time is updated only on first and last access as expected.
 TEST_F(TestQuotaManager, OpenClientDirectory_Ongoing_OriginDirectoryExists) {
+  auto testOriginMetadata = GetTestOriginMetadata();
+
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
@@ -666,6 +739,15 @@
   const auto saveOriginAccessTimeCountInternalBefore =
       SaveOriginAccessTimeCountInternal();
 
+  auto originStateMetadataBefore = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataBefore);
+  ASSERT_FALSE(originStateMetadataBefore->mAccessed);
+
+  auto directoryMetadataHeaderBefore =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_TRUE(directoryMetadataHeaderBefore);
+  ASSERT_FALSE(directoryMetadataHeaderBefore->mAccessed);
+
   PerformOnBackgroundThread([saveOriginAccessTimeCountBefore,
                              saveOriginAccessTimeCountInternalBefore]() {
     auto testClientMetadata = GetTestClientMetadata();
@@ -798,6 +880,15 @@
                 saveOriginAccessTimeCountInternalBefore,
             2u);
 
+  auto originStateMetadataAfter = GetOriginStateMetadata(testOriginMetadata);
+  ASSERT_TRUE(originStateMetadataAfter);
+  ASSERT_TRUE(originStateMetadataAfter->mAccessed);
+
+  auto directoryMetadataHeaderAfter =
+      LoadDirectoryMetadataHeader(testOriginMetadata);
+  ASSERT_TRUE(directoryMetadataHeaderAfter);
+  ASSERT_TRUE(directoryMetadataHeaderAfter->mAccessed);
+
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
 
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
@@ -3122,6 +3213,7 @@
 
   auto originStateMetadata = maybeOriginStateMetadata.extract();
   ASSERT_GT(originStateMetadata.mLastAccessTime, 0);
+  ASSERT_FALSE(originStateMetadata.mAccessed);
   ASSERT_FALSE(originStateMetadata.mPersisted);
 
   ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
